src/rust/engine/fs/src/directory_tests.rs

```diff
@@ -0,0 +1,372 @@
+      path: Path::new("dir/self_but_oh_so_obtusely"),
+      target: Path::new("self/self/self/self/self_obtusely/parent/dir/parent/dir/parent/dir/self"),
+    },
+    TypedPath::File {
+      path: Path::new("dir/file.txt"),
+      is_executable: false,
+    },
+  ]);
+  assert_entry_is_some(&tree, "dir/self_but_oh_so_obtusely");
+  assert_entry_is_some(&tree, "dir/self_but_oh_so_obtusely/file.txt");
+}
+
+fn assert_walk(tree: &DigestTrie, expected_filenames: Vec<String>, expected_dirnames: Vec<String>) {
+  let mut filenames = Vec::new();
+  let mut dirnames = Vec::new();
+  tree.walk(
+    crate::SymlinkBehavior::Oblivious,
+    &mut |path, entry| match entry {
+      Entry::Symlink(_) => panic!("But we're oblivious!"),
+      Entry::Directory(_) => dirnames.push(path.to_path_buf()),
+      Entry::File(_) => filenames.push(path.to_path_buf()),
+    },
+  );
+  assert_eq!(
+    filenames,
+    expected_filenames
+      .iter()
+      .map(|s| PathBuf::from(s))
+      .collect::<Vec<_>>()
+  );
+  assert_eq!(
+    dirnames,
+    expected_dirnames
+      .iter()
+      .map(|s| PathBuf::from(s))
+      .collect::<Vec<_>>()
+  );
+}
+
+#[test]
+fn walk_simple() {
+  let tree = make_tree(vec![
+    TypedPath::File {
+      path: Path::new("file.txt"),
+      is_executable: false,
+    },
+    TypedPath::Link {
+      path: Path::new("symlink"),
+      target: Path::new("file.txt"),
+    },
+    TypedPath::Link {
+      path: Path::new("relsymlink"),
+      target: Path::new("./file.txt"),
+    },
+    TypedPath::Link {
+      path: Path::new("a/symlink"),
+      target: Path::new("../file.txt"),
+    },
+    TypedPath::Link {
+      path: Path::new("a/b/symlink"),
+      target: Path::new("../../file.txt"),
+    },
+  ]);
+  assert_walk(
+    &tree,
+    vec![
+      "a/b/symlink".to_string(),
+      "a/symlink".to_string(),
+      "file.txt".to_string(),
+      "relsymlink".to_string(),
+      "symlink".to_string(),
+    ],
+    vec!["".to_string(), "a".to_string(), "a/b".to_string()],
+  );
+}
+
+#[test]
+fn walk_too_many_links_rootdir() {
+  let tree = make_tree(vec![
+    TypedPath::File {
+      path: Path::new("file.txt"),
+      is_executable: false,
+    },
+    TypedPath::Link {
+      path: Path::new("self"),
+      target: Path::new("."),
+    },
+  ]);
+  assert_walk(
+    &tree,
+    (0..MAX_LINK_DEPTH)
+      .into_iter()
+      .map(|n| ("self/".repeat(n.into()) + "file.txt"))
+      .collect::<Vec<_>>(),
+    vec!["".to_string()],
+  );
+}
+
+#[test]
+fn walk_too_many_links_subdir() {
+  let tree = make_tree(vec![
+    TypedPath::File {
+      path: Path::new("a/file.txt"),
+      is_executable: false,
+    },
+    TypedPath::Link {
+      path: Path::new("a/self"),
+      target: Path::new("."),
+    },
+  ]);
+  assert_walk(
+    &tree,
+    (0..MAX_LINK_DEPTH)
+      .into_iter()
+      .map(|n| ("a/".to_string() + &"self/".repeat(n.into()) + "file.txt"))
+      .collect::<Vec<_>>(),
+    vec!["".to_string(), "a".to_string()],
```

Comment:
Oh yeah @stuhood here's what I'm talking about. `a/self/file.txt` is a file (entry is for `a/file.txt`, but the only directory entries are root and `a`. 

I think conceptually it makes sense. There is no entry for `a/self`. But an interesting quirk.